### ISSUE TITLE
The value of the Type of transaction in Transactions page is not being displayed as intended

### ISSUE THREAD

**Issue by zbig01** [openemr/openemr#1825]
A minor bug is noted when a Patient Transaction is created on the Transactions page, the value of the Type of transaction is not being displayed as it was in the past. The Type is displayed as LBTref, LBTptreq etc with an exclamation icon as a superscript.

![pa_transactions_1](https://user-images.githubusercontent.com/12654295/44833747-ed2ee700-abe3-11e8-85c8-891319696577.png)

This is caused by :

`<?php echo generate_display_field(['data_type' => 1, 'list_id' => 'transactions'], $item['title']); ?>`

on line 122 of `interface/patient_file/transaction/transactions.php`

Upon looking at the code for the `generate_display_field` function  in ` library/options.inc.php` we see the query on lines 2028 and 2028

```
$lrow = sqlQuery("SELECT title FROM list_options " .
        "WHERE list_id = ? AND option_id = ? AND activity = 1", array($list_id,$currvalue));
```
        
is seeking the value for title in table `list_options`.

These values that were introduced by @sunsetsystems committed on Jul 8, 2015 1 parent f4b60b9 commit 4ea4311875fcb9a7a5db98e109c29167a6e2242a in `sql/database.sql`

```
INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('lists', 'transactions', 'Layout-Based Transaction Forms', 9, 0);
INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTref'  , 'Referral'         , 10, 0);
INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTptreq', 'Patient Request'  , 20, 0);
INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTphreq', 'Physician Request', 30, 0);
INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTlegal', 'Legal'            , 40, 0);
INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTbill' , 'Billing'          , 50, 0);
```

were removed by @sunsetsystems authored and bradymiller committed on Sep 15, 2017 1 parent 5b149cd commit db79047421807dcfa6c4826489e528c6ea86201b

```
-- INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('lists', 'transactions', 'Layout-Based Transaction Forms', 9, 0);
-- INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTptreq', 'Patient Request'  , 20, 0);
-- INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTphreq', 'Physician Request', 30, 0);
-- INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTlegal', 'Legal'            , 40, 0);
-- INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES ('transactions', 'LBTbill' , 'Billing'          , 50, 0);
```

causing this line of code in `function generate_display_field` to be executed

```
// If match is not found in main and backup lists, return the key with exclamation mark
if ($s == '') {
    $s = nl2br(htmlspecialchars($currvalue, ENT_NOQUOTES)).
    '<sup> <i class="fa fas fa-exclamation-circle ml-1"></i></sup>';
```

Thus leading to the altered and less meaningful display of the type of transaction.

**Comment by bradymiller** [comment]
Does changing title to option_id in the sql query select and then changing:
```$item['title']```
to
```$item['option_id']```
fix the generate_display_field() call

**Comment by bradymiller** [comment]
Actually forget about changing the sql query. That is fine. I think just changing to ```$item['option_id']``` should fix this

**Comment by zbig01** [comment]
The issue is in `library/options.inc.php`
```
$lrow = sqlQuery("SELECT title FROM list_options " .
        "WHERE list_id = ? AND option_id = ? AND activity = 1", array($list_id,$currvalue));
```
the sql query is run on table list_options, it will always fail  as the values that it seeks have been removed from the table

BTW  changing:
`$item['title']`
to
`$item['option_id']`
in the `generate_display_field()` call in `interface/patient_file/transaction/transactions.php` causes only the exclamation mark to show

**Comment by bradymiller** [comment]
hi @zbig01 ,
Thanks for pointing this out. Looks like the titles got migrated to the layout_group_properties tables in the big layouts project that @sunsetsystems did. Looks like we need a function like collectLayoutNames() that is in the edit_layout.php script to grab the titles.
-brady

**Comment by zbig01** [comment]
I fixed the issue by making 2 changes:

1.  `library/options.inc.php` - added to `function generate_display_field`
```
elseif ($data_type == 99) {
        $lrow = sqlQuery("SELECT grp_title FROM layout_group_properties " .
        "WHERE grp_mapping = ? AND grp_form_id = ? ", array($list_id,$currvalue));
        $s = htmlspecialchars(xl_list_label($lrow['grp_title']), ENT_NOQUOTES);
        
        // If match is not found  return the key with exclamation mark
        if ($s == '') {
            $s = nl2br(htmlspecialchars($currvalue, ENT_NOQUOTES)).
            '<sup> <i class="fa fas fa-exclamation-circle ml-1"></i></sup>';
        }
    }
```


I used `$data_type == 99` because I did not want to mess up any other place where `$data_type == 1` was being used

2. modified line 122 of interface/patient_file/transaction/transactions.php
`<?php echo generate_display_field(['data_type' => 99, 'list_id' => 'Transactions'], $item['title']); ?>`


![pa_transactions_2](https://user-images.githubusercontent.com/12654295/44902486-fc369780-acbe-11e8-8974-222be467bdba.png)

**Comment by bradymiller** [comment]
very nice!

Check out this function, getListItemTitle(), that we use for list titles (generate_display_field is the old way of doing this):
https://github.com/openemr/openemr/blob/master/library/options.inc.php#L3806-L3817

Recommend placing your code in a new function called getLayoutTitle() in options.inc.php, and then calling that instead. Then don't need to hack out a data_type.

-brady

**Comment by bradymiller** [comment]
So you don't get weighed down with this on your ui stuff, just let me know if you want me to do this (I could make this quick fix); I'd be happy to :)

**Comment by zbig01** [comment]
@bradymiller 

minor bug fix in patient transactions #1830
